viite-backend/viite-main/src/main/scala/fi/vaylavirasto/viite/dynamicnetwork/DynamicRoadNetworkService.scala

```diff
@@ -0,0 +1,564 @@
+
+        val linearLocationsWithOldLinkId = linearLocations.filter(_.linkId == oldLinkId)
+        val roadAddressedLinkLength = GeometryUtils.scaleToThreeDigits(linearLocationsWithOldLinkId.map(_.endMValue).max - linearLocationsWithOldLinkId.map(_.startMValue).min)
+
+        // check that there are no "partial" changes to road addressed links, i.e. only part of the link changes and the other part has no changes applied to it.
+        if (lengthOfChange != roadAddressedLinkLength) {
+          val allChangesWithOldLinkId = tiekamuRoadLinkChanges.filter(_.oldLinkId == oldLinkId)
+          val combinedLengthOfChanges = GeometryUtils.scaleToThreeDigits(allChangesWithOldLinkId.map(och => och.oldEndM - och.oldStartM).sum)
+          if (combinedLengthOfChanges != roadAddressedLinkLength)
+            tiekamuRoadLinkChangeErrors += TiekamuRoadLinkChangeError("No partial changes allowed. The old link needs to have changes applied to the whole length of the old link", change, getMetaData(change, linearLocations))
+        }
+        else if (newlinkIdChangesLength != GeometryUtils.scaleToThreeDigits(newLinkKGV.length)) {
+          tiekamuRoadLinkChangeErrors += TiekamuRoadLinkChangeError("No partial changes allowed. The new link needs to have changes applied to the whole length of the new link", change, getMetaData(change, linearLocations))
+        }
+
+        // if there are combined links (A + B = C)
+        else if (otherChangesWithSameNewLinkId.nonEmpty) {
+          val oldLinkIds = otherChangesWithSameNewLinkId.map(_.oldLinkId)
+          val oldLinearLocations = linearLocations.filter(ll => oldLinkIds.contains(ll.linkId))
+          val roadways = roadwayDAO.fetchAllByRoadwayNumbers(oldLinearLocations.map(_.roadwayNumber).toSet)
+          // group the roadways with roadNumber, roadPartNumber and track
+          val roadGroups = roadways.groupBy(rw => (rw.roadNumber, rw.roadPartNumber, rw.track))
+          // if there are change infos that combine two or more links but the road address is not homogeneous between those merging links then its an error
+          val isRoadAddressHomogeneous = checkRoadAddressHomogeneity(roadGroups)
+          val changesInSameRoadway = roadGroups.head._2.map(_.roadwayNumber).distinct.size == 1
+          if (!isRoadAddressHomogeneous)
+            tiekamuRoadLinkChangeErrors += TiekamuRoadLinkChangeError("Two or more links with non homogeneous road addresses (road number, road part number, track) cannot merge together", change, getMetaData(change, linearLocations))
+          else if (isRoadAddressHomogeneous && changesInSameRoadway) {
+            if (!checkRoadAddressContinuityForSingleRoadway(oldLinearLocations, change))
+              tiekamuRoadLinkChangeErrors += TiekamuRoadLinkChangeError("Road address not continuous, cannot merge links together", change, getMetaData(change, linearLocations))
+          }
+          else if (isRoadAddressHomogeneous && !changesInSameRoadway) {
```

Comment:
r. 451-455 taitaa olla copy-paste-duplikaatti?

Reply:
Ei oo duplikaatti, siinä on ! merkki toisessa if lauseessa

&& !changesInSameRoadway

vs

 && changesInSameRoadway
 
 ja niiden sisällä olevat iffit on myös eri

Reply:
Tou! Niinpäs onkii. 
No mut sit mä ehkä huitoisin casejen (tai ylipäänsä sisennysrakenteen tai whatnot) käyttöön
Ni sit toi eri ehtojen suhde toisiinsa tulee selkeemmin näkyviin.
Ei ehkä printtaudu tähän nätisti - mutta esim. tällee:

          isRoadAddressHomogeneous match {
            case false => tiekamuRoadLinkChangeErrors += TiekamuRoadLinkChangeError("Two or more links with non homogeneous road addresses (road number, road part number, track) cannot merge together", change, getMetaData(change, linearLocations))
            case true  => { 
              changesInSameRoadway match {
                case true  => if (!checkRoadAddressContinuityForSingleRoadway(oldLinearLocations, change))
                           tiekamuRoadLinkChangeErrors += TiekamuRoadLinkChangeError("Road address not continuous, cannot merge links together", change, getMetaData(change, linearLocations))
                case false => if (!checkRoadAddressContinuityBetweenRoadways(oldLinearLocations, change))
                           tiekamuRoadLinkChangeErrors += TiekamuRoadLinkChangeError("Road address not continuous, cannot merge links together", change, getMetaData(change, linearLocations))
              }
            }
          }
          
          (jos ton copy-pasteaa, niin kannattanee vielä varmistaa, että laitoin ehtojen truet ja falset oikein päin, ettei mee toiminnallisuus rikki)